ncp: Miscellaneous updates

darconeous opened the pull request:
This pull request contains the following changes:

* 3e0d503 spinel: Rename `PROP_PHY_PROMISCUOUS_MODE` to `PROP_PHY_MODE`. `PROP_PHY_MODE` is now an enumeration of the following values:    
    * `PHY_MODE_NORMAL`: Normal filtering is in place.
    * `PHY_MODE_PROMISCUOUS`: Only PANID filtering is in place.
    * `PHY_MODE_MONITOR`: All channel traffic is passed up from the PHY.
* 7c1cee5 spinel: Add more status codes and related descriptions.
* c041518 spinel: Add support for fetching the thread network data.
* 9fab9e5 ncp: Add support for the `STREAM_NET_INSECURE` data stream.
* 944e44a ncp: Support returning LQI for scan results.
* edf3462 spinel: Add missing definition of `spinel_prop_key_to_cstr()`.
* a1421ae spinel: Add new thread properties.
     *  `PROP_THREAD_ON_MESH_NETS` : For advertising this device as a border router with an on-mesh prefix.
     *  `PROP_THREAD_LOCAL_ROUTES` : For advertising more specific routes that this node can provide. (Was previously
        `PROP_IPV6_EXT_ROUTE_TABLE`)
     *  `PROP_THREAD_ASSISTING_PORTS` : A list of port numbers that are allowed to be passed via `PROP_STREAM_NET_INSECURE` when received without security. Empty when this device is not an assisting device.
* 0b4f852 ncp: Fix for `PROP_FREQ`
    * The 2.4GHz range was being computed incorrectly. We were also returning the value of the pointer to the frequency. instead of the frequency.
* 084d987 ncp: Code style update.

turon commented:
Hi RQ, I've generally seen the term PROMISCUOUS to mean no filtering at all: neither PANID nor short address nor extended address. I'd suggest the following naming:
* PHY_FILTER_MODE_NORMAL: Normal device filtering is in place, e.g. address and PANID.
* PHY_FILTER_MODE_NETWORK: Only PANID filtering is in place.
* PHY_FILTER_MODE_PROMISCUOUS: All channel traffic is passed up from the PHY.

darconeous commented:
@turon: Wi-Fi uses the model I implemented here: *promiscuous* mode means network-only, whereas *monitor* is the big fat firehose. This convention is [well established](http://security.stackexchange.com/questions/36997/what-is-the-difference-between-promiscuous-and-monitor-mode-in-wireless-networks), and I think it has utility to wireless network debugging in general—which is why I have adopted that convention here.

In cases where the underlying hardware doesn't support the concept of filtering on only the PANID versus no filtering at all, changing the mode to `PHY_MODE_PROMISCUOUS` would result in `PHY_MODE_MONITOR` being set instead. (Or, if we wanted to, we could implement that filtering in software)

turon commented:
@darconeous I'm used to the IEEE 802.15.4-2006 definition, which in section 7.5.6.5 Promiscuous mode
defines it as simply passing all packets from the PHY (which doesn't define PANID) to the next higher layer.  otPlatRadio is essentially a PHY interface.  I'm fine choosing terms that work across multiple radio standards, but don't think we should choose terminology that redefines the meaning from a IEEE 802.15.4 perspective.

darconeous commented:
I'm trying to keep the `PROP_PHY_*` properties as general purpose as possible. As such I tend to prefer the general widespread convention rather than force the limited 802.15.4 definitions on everything.

Alternatively, I'd be happy to define a `PHY_MODE_802_15_4_PROMISCUOUS` that is defined to be the same numerical value as `PHY_MODE_MONITOR`.

Of course, if there is consensus among the maintainers to follow the 802.15.4 convention, I'll do that. 

On an unrelated note, I'm considering moving this property from the `PHY` section to the `MAC` section.

jwhui commented:
As I understand it, spinel was intended to incorporate link technologies beyond just IEEE 802.15.4.  If this is true, then I think spinel should use the definition that is more commonly used across different link technologies.

I agree that this property should be moved to the MAC section.

darconeous commented:
For what it's worth, there is always `CAP_802_15_4_PIB`, which maps the entire 802.15.4 PIB space to the properties between `PROP_15_4_PIB__BEGIN` (1024) and `PROP_15_4_PIB__END` (1280).

*macPromiscuousMode* maps to the 802.15.4 PIB identifier `0x51`. Thus, if a device has the `CAP_802_15_4_PIB` capability, it can use the property `PROP_15_4_PIB_MAC_PROMISCUOUS_MODE` (property id of 1105), which would behave as intended.

I doubt many NCPs will implement that optional capability, but it is nonetheless defined in the protocol.

turon commented:
@darconeous Sounds good.  I support the more general link terms.  Having your suggested enum alias should be enough to clear up confusion by those more accustomed to 15.4 terminology.  This discussion taught me something, and broke me free of a 15.4-centric view, thanks!

darconeous commented:
Oops, forgot that alias. Will add.

darconeous commented:
Ok, [added the definition of `SPINEL_MAC_FILTER_MODE_15_4_PROMISCUOUS`](https://github.com/openthread/openthread/pull/88/files#diff-7acc5ee1cf7912cf8cc670e77c688d53R164).

darconeous commented:
Somehow one of my `[NOT-FOR-UPSTREAM]` commits made it into the pull request and caused it to screw up the build checks. I've removed the commit and rebased to master.

jwhui commented:
LGTM 👍

darconeous commented:
Wow, didn't realize that all the time I spent breaking out all of those changes into individual commits was all for nothing. Everything was collapsed into a single commit! :/

This git repository works differently than any other git repository I've worked with on Github.
